pkg/controller/master/nodedrain/nodedrain_controller.go

```diff
@@ -0,0 +1,293 @@
+	}
+
+	var possibleVolumeNames []string
+
+	for pvcName, replicaState := range volumeMap {
+		if replicaState.healthy <= 1 {
+			for _, v := range replicaState.replicas {
+				// last started replica is on the current node
+				if v.Status.Started && v.Spec.NodeID == node.Name {
+					possibleVolumeNames = append(possibleVolumeNames, pvcName)
+				}
+			}
+		}
+	}
+
+	volList := make([]*longhornv1.Volume, 0, len(possibleVolumeNames))
+	for _, v := range possibleVolumeNames {
+		vObj, err := ndc.longhornVolumeCache.Get(util.LonghornSystemNamespaceName, v)
+		if err != nil {
+			return nil, err
+		}
+		volList = append(volList, vObj)
+	}
+
+	return volList, nil
+}
+
+// findAndListVM is called by action handler to leverage caches to find unhealthy VM's impacted by the migration
+func (ndc *ControllerHandler) FindAndListVM(node *corev1.Node) ([]string, error) {
+	shutdownVMs, err := ndc.listVMI(node)
+	if err != nil {
+		return nil, fmt.Errorf("error listing VMI: %v", err)
+	}
+	impactedVMDetails := make([]string, 0, len(shutdownVMs))
+	for _, v := range shutdownVMs {
+		vmName, err := findVM(v)
+		if err != nil {
+			return nil, err
+		}
+		impactedVMDetails = append(impactedVMDetails, fmt.Sprintf("VM %s in namespace %s is unhealthy", vmName, v.Namespace))
+	}
+	return impactedVMDetails, nil
+}
+
+func ActionHelper(nodeCache ctlcorev1.NodeCache, virtualMachineInstanceCache kubevirtv1.VirtualMachineInstanceCache,
+	longhornVolumeCache longhornv1beta1.VolumeCache, longhornReplicaCache longhornv1beta1.ReplicaCache) *ControllerHandler {
+	return &ControllerHandler{
+		nodeCache:                   nodeCache,
+		virtualMachineInstanceCache: virtualMachineInstanceCache,
+		longhornVolumeCache:         longhornVolumeCache,
+		longhornReplicaCache:        longhornReplicaCache,
+	}
+}
+
+// isLastNode will check if there are other nodes in cluster which are not in maintenance mode
+func (ndc *ControllerHandler) meetsControlPlaneRequirements(node *corev1.Node) (bool, error) {
+
+	_, cpLabelOK := node.Labels["node-role.kubernetes.io/control-plane"]
+	if !cpLabelOK { // not a controlplane node. no further action needed
+		return true, nil
+	}
+
+	cpSelector := labels.SelectorFromSet(map[string]string{
+		"node-role.kubernetes.io/control-plane": "true",
+	})
+	nodeList, err := ndc.nodeCache.List(cpSelector)
+	if err != nil {
+		return false, fmt.Errorf("error listing nodes matching selector %s: %v", cpSelector.String(), err)
+	}
+
+	// only controlplane node which we are trying to place into maintenance
+	if len(nodeList) == 1 {
+		return false, nil
+	}
+
+	var availableNodes int
+	for _, v := range nodeList {
+		_, ok := v.Annotations[ctlnode.MaintainStatusAnnotationKey]
+		logrus.Infof("nodeName: %s,  annotation present: %v\n", v.Name, ok)
+		if !ok {
+			availableNodes++
+		}
+	}
+
+	// only 1 node can be placed in maintenance mode at a time. All cp nodes should be available
+	// for a node to be placed in maintenance mode
+	return availableNodes == 3, nil
```

Comment:
3 can be replaced by https://github.com/harvester/harvester/blob/6fd9a5bb3b99a329101ec70f657a30e24906035f/pkg/controller/master/node/promote_controller.go#L42

Reply:
👍